Fix set operators: UNION, EXCEPT and INTERSECT

gmantele opened the pull request:
A proposal to fix the grammar of the set operators and their description.

Resolves #20

gmantele committed:
Fix grammar of `UNION`, `EXCEPT` and `INTERSECT`.

A table name or join is no longer directly allowed as left or right
operand of a set operator (on the contrary to what SQL92 seems to allow ;
the SQL92 rules are not supported by the nowadays major DBMS).

Without parentheses, only a `SELECT` without `WITH`, `ORDER BY` and `OFFSET`
clauses is possible. Note that any set operations are still possible in such
limited `SELECT` query.

With parentheses, a complete `SELECT` query (so, with `WITH`, `ORDER BY`
and `OFFSET` clauses) is possible.

Set operations are still evaluated with the following precedence:

1. Expressions in parentheses
2. The `INTERSECT` operation
3. `EXCEPT` and `UNION` (from left to right)

gmantele committed:
Fix `INTERSECT` description and add clarifications on operands and precedence of all set operators

gmantele commented:
The grammar fix proposed here should allow the following ADQL queries:

* **Simple `SELECT`:**
  ```sql
      SELECT hip FROM hipparcos WHERE hip <= 10
  UNION
      SELECT hip FROM hipparcos WHERE hip > 120406
  ORDER BY 1 DESC
  ```

* **`SELECT TOP ...`:**
  ```sql
      SELECT hip FROM hipparcos WHERE hip <= 10
  UNION
      SELECT TOP 10 hip FROM hipparcos WHERE hip > 10
  ORDER BY 1 DESC
  ```

* **`SELECT TOP ... ORDER BY ...`:**
  ```sql
      SELECT hip FROM hipparcos WHERE hip <= 10
  UNION
      (SELECT TOP 10 hip FROM hipparcos WHERE hip > 10 ORDER BY 1)
  ORDER BY 1 DESC
  ```

* **`INTERSECT`:**
  ```sql
      SELECT hip FROM hipparcos WHERE hip <= 10
  INTERSECT
      SELECT hip FROM hipparcos WHERE hip <= 10 AND mod(hip, 2) = 0
  ORDER BY 1 DESC
  ```

* **`EXCEPT`:**
  ```sql
      SELECT hip FROM hipparcos WHERE hip <= 10
  EXCEPT
      SELECT hip FROM hipparcos WHERE hip <= 10 AND mod(hip, 2) = 0
  ORDER BY 1 DESC
  ```

* **Precedence `UNION` vs `INTERSECT`:**
  ```sql
      SELECT hip FROM hipparcos WHERE hip <= 10
  UNION
      SELECT hip FROM hipparcos WHERE hip > 120406
  INTERSECT
      SELECT hip FROM hipparcos WHERE (hip <= 10 OR hip > 120406) AND mod(hip, 2) = 0
  ORDER BY 1 DESC
  ```

* **Precedence with parentheses:**
  ```sql
  (
      SELECT hip FROM hipparcos WHERE hip <= 10
  UNION
      SELECT hip FROM hipparcos WHERE hip > 120406
  )
  INTERSECT
      SELECT hip FROM hipparcos WHERE (hip <= 10 OR hip > 120406) AND mod(hip, 2) = 0
  ORDER BY 1 DESC
  ```

* **With CTE:**
  ```sql
  WITH tt AS (SELECT hip FROM hipparcos WHERE hip > 120406)
      SELECT hip FROM hipparcos WHERE hip <= 10
  UNION
      SELECT hip FROM tt
  ORDER BY 1 DESC
  ```

* **Embedded CTE:**
  ```sql
      SELECT hip FROM hipparcos WHERE hip <= 10
  UNION
      (
          WITH tt AS (SELECT hip FROM hipparcos WHERE hip > 120406)
          SELECT hip FROM tt
      )
  ORDER BY 1 DESC
  ```

I tested their SQL translation with PostgreSQL, MySQL and MS-SQLServer.
Although some syntactic changes were required depending on the DBMS, all of them
pass.

=> Tell me if I forgot some important cases.

gmantele commented:
Some notes about the MySQL and MS-SQLServer:

`INTERSECT` and `EXCEPT` do not exist in MySQL, but it is possible to get the
same result with joins.

Embedded CTE with MS-SQLServer does not work. The only hack I found is to put
all CTEs at the root level. Let me know if this hack will not be enough in
some cases.

msdemlei commented:
On Fri, Oct 23, 2020 at 10:34:23AM -0700, Grégory Mantelet wrote:
> The grammar fix proposed here should allow the following ADQL queries:

I think I agree with all of the queries except this one:

> * **Embedded CTE:**
>   ```sql
>       SELECT hip FROM hipparcos WHERE hip <= 10
>   UNION
>       (
>           WITH tt AS (SELECT hip FROM hipparcos WHERE hip > 120406)
>           SELECT hip FROM tt
>       )
>   ORDER BY 1 DESC
>   ```

Is there an overriding reason to allow CTEs in operands of set
operators?  I cannot see how that would raise the expressiveness of
the language (and I think your parallel comment about pulling CTEs to
the top leven in SQLServer says about as much).

I see a slight advantage in that you can stick arbitrary queries into
such arguments, which might be nice when composing queries out of
pre-fabricated fragments, but then the cut-and-paste effort of taking
the inner with_queries and sticking them to the top level doesn't
seem to be prohibitive, and you get queries that are better
understandable because there are fewer places you might have to check
for table expressions.

The reason I'd rather not have these "embedded" CTEs is that table
reolution (i.e., figuring out the table definition for a given table
reference) is hard enough already.  When CTEs are only allowed at the
top level, it doesn't get much harder (I was actually positively
surprised how little code was necessary for that).  Allowing CTEs in
extra positions will make that a good deal more complicated.

So: I'd say embedded CTEs are little gain for noticable pain, and
hence I'd say we shouldn't have them (unless someone finds a stronger
rationale than the cut-and-pastee case I'm making above).

In grammar terms, I think that would mean introducing a:

```
<select_expression> ::= 
       <query_set_expression>
        [ <order_by_clause> ]
        [ <offset_clause> ]
```

With that, we'd have:

```
<query_specification> ::=
        WITH <with_query> {, <with_query>} ...
        <select_expression>

<query_set_primary> ::=
        <select_query>
        | <left_paren> <select_expression> <right_parent>

<query_expression> ::=
        <select_expression>
        | <joined_table>
```

Incidentally, even if there's actually a case for CTEs in set
operands, I'm pretty sure we wouldn't want the CTEs in a normal
subquery (which is what query_expression is used for) -- or do we?

gmantele commented:
I agree, I do not see the point of having CTEs inside a set's operand. I added this query example because there is no reason why it should not be possible considering that CTEs are currently allowed in "normal" subqueries : this way, it is consistent. Besides, it works like that in most DBMS.

However, I admit I have not yet started to implement the set operators in my ADQL parser (but I will this week...and before merging this PullRequest for sure), so I am now very sensible to the point @msdemlei raises about the table resolution : table resolution will be indeed more complicated than it is now (and it is already a bit challenging).

As @msdemlei , I agree that CTEs should be allowed only at the top level query and not any more in subqueries or set's operands. This will help parser development but also it will make ADQL queries more readable (all CTEs will always be declared at only one place: right before the main query).

Me too, I can not find any example which will make "embedded" CTEs different than having them declared at the top level.
Does anybody else have any example?

gmantele commented:
About the "embedded" CTEs, I do not have to change anything in my parser to make it work, even for the table resolution.
So, either solution (allowing them, or not) is fine with me, on an implementation point of view.

Now, because it makes appararently no difference whether the CTEs are embedded or not, and because it would make ADQL queries more clear, I propose that in ADQL(-2.1) CTE are only accepted at the root level.

So, to sum up with examples:

```sql
    SELECT ra, dec FROM myTable
UNION
    (WITH cteTable AS (SELECT ra, dec FROM anotherTable)
     SELECT * FROM cteTable)
```

won't be accepted in ADQL, but the following will be:

```sql
WITH cteTable AS (SELECT ra, dec FROM anotherTable)
    SELECT ra, dec FROM myTable
UNION
    SELECT * FROM cteTable
```

This way, if later we discover there are use-cases where embedded CTEs are the only solution, we can easily extend the ADQL grammar without breaking anything.

I change the grammar right now. If any opposition, speak now (or before this PullRequest is accepted/merged).

gmantele committed:
Grammar Fix: Forbid CTEs in set operations' operands.
They are allowed only at the main level.

gmantele committed:
Wording Fix: Forbid CTEs in set operations' operands.
They are allowed only at the main level.

gmantele committed:
Update query syntax description

gmantele commented:
I moved the section about the `WITH` clause (or CTEs), just before the set operators so that speaking of CTEs in the set operators section makes sense.

I also added the following sentence in the WITH section:

> For clarity reason, common table expressions can be defined only in the main query. They are not allowed in sub-queries.

jd-au commented on ADQL.tex:
```diff
@@ -2636,7 +2708,7 @@ \subsubsection{UNION}
 For a UNION operation to be valid in ADQL, the following criteria MUST be met:
 \begin{itemize}
-    \item the two queries MUST result in the same number of columns
+    \item the two queries MUST result in the same number of columns.
```
I am not sure why you have added a full stop to the end of each dot point in this section. Having no full stop seems to be more widely accepted unless each dot point is a sentence starting with a capital letter.

gmantele replied:
It was just because the next item ends with a full stop...it was for consistency. If you tell me it is better without this full stop, I remove it ;-)

gmantele replied:
Done

jd-au commented on ADQL.tex:
```diff
@@ -2717,72 +2789,78 @@ \subsubsection{INTERSECT}
-An ADQL service implementation MAY include support for the following optional
-support for common table expressions:
+Operands of any of the set operators can only be \verb:SELECT: queries.
+Unless if within parentheses, such query can not use any \verb:ORDER BY: or
```
Suggest `Unless within parentheses, such queries`

gmantele replied:
Ok. I'll fix it.

gmantele committed:
Fix typo in Set operation's "Operands" subsection

gmantele committed:
Fix typo in ending some list items

gmantele commented:
It's time for whoever is ready to review this Pull Request.
According to me, this Pull Request and its corresponding issue can now be closed.

molinaro-m reviewed (approved):
I approve the changes in the text and typo fixes.

Zarquan reviewed (approved):
Looks good to me :+1:
